adal/src/main/java/com/microsoft/aad/adal/BrokerProxy.java

```diff
@@ -558,34 +567,154 @@ private boolean hasSupportToAddUserThroughBroker(final String brokerPackageName)
 
     private boolean verifySignature(final String brokerPackageName) {
         try {
-            PackageInfo info = mContext.getPackageManager().getPackageInfo(brokerPackageName,
-                    PackageManager.GET_SIGNATURES);
-
-            if (info != null && info.signatures != null) {
-                // Broker App can be signed with multiple certificates. It will
-                // look all of them until it finds the correct one for ADAL
-                // broker.
-                for (Signature signature : info.signatures) {
-                    MessageDigest md = MessageDigest.getInstance("SHA");
-                    md.update(signature.toByteArray());
-                    String tag = Base64.encodeToString(md.digest(), Base64.NO_WRAP);
-
-                    // Company portal(Intune) app and Azure authenticator app
-                    // have authenticator.
-                    if (tag.equals(mBrokerTag)
-                            || tag.equals(AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_SIGNATURE)) {
-                        return true;
-                    }
-                }
+            // Read all the certificates associated with the package name. In higher version of
+            // android sdk, package manager will only returned the cert that is used to sign the
+            // APK. Even a cert is claimed to be issued by another certificates, sdk will return
+            // the signed. However, for the lower version of android, it will return all the
+            // certs in the chain. We need to verify
+            final List<X509Certificate> certs = readCertDataForBrokerApp(brokerPackageName);
+
+            // If there is only one cert returned, no need to perform certificate chain validation.
+            // It should be 1) only one self signed cert exists 2) The OS iteslf is only returning
+            // back the signing cert. We only need to validate the signing cert signature matches
+            // what we hardcoded.
+            if (certs.size() == 1) {
+                return verifySignatureHash(certs.get(0));
             }
+
+            // Verify the certificate chain is chained correctly.
+            final X509Certificate singingCert = getSignerCertAndVerifyCertificateChain(certs);
+            return verifySignatureHash(singingCert);
         } catch (NameNotFoundException e) {
             Logger.e(TAG, "Broker related package does not exist", "", ADALError.BROKER_PACKAGE_NAME_NOT_FOUND);
         } catch (NoSuchAlgorithmException e) {
             Logger.e(TAG, "Digest SHA algorithm does not exists", "", ADALError.DEVICE_NO_SUCH_ALGORITHM);
+        } catch (final AuthenticationException | IOException | GeneralSecurityException e) {
+            Logger.e(TAG, e.getMessage(), "", ADALError.BROKER_VERIFICATION_FAILED, e);
         }
+
         return false;
     }
 
+    boolean verifySignatureHash(final X509Certificate signerCert) throws NoSuchAlgorithmException,
```

Comment:
this function never returns false, you should reconsider refactoring it to be void and in caller above return true if it does not throw